Library/Homebrew/dev-cmd/extract.rb

```diff
@@ -0,0 +1,153 @@
+#:  * `extract` [`--force`] <formula> `--tap=`<tap> [`--version=`<version>]:
+#:    Looks through repository history to find the <version> of <formula> and
+#:    creates a copy in <tap>/Formula/<formula>@<version>.rb. If the tap is
+#:    not installed yet, attempts to install/clone the tap before continuing.
+#:    A tap must be passed through `--tap` in order for `extract` to work.
+#:
+#:    If `--force` is passed, the file at the destination will be overwritten
+#:    if it already exists. Otherwise, existing files will be preserved.
+#:
+#:    If an argument is passed through `--version`, <version> of <formula>
+#:    will be extracted and placed in the destination tap. Otherwise, the most
+#:    recent version that can be found will be used.
+
+require "utils/git"
+require "formulary"
+require "tap"
+
+class BottleSpecification
+  def method_missing(m, *_args, &_block)
+    if [:sha1, :md5].include?(m)
+      opoo "Formula has unknown or deprecated stanza: #{m}" if ARGV.debug?
+    else
+      super
+    end
+  end
+end
+
+class Module
+  def method_missing(m, *_args, &_block)
+    if [:sha1, :md5].include?(m)
+      opoo "Formula has unknown or deprecated stanza: #{m}" if ARGV.debug?
+    else
+      super
+    end
+  end
+end
+
+class DependencyCollector
+  def parse_symbol_spec(spec, tags)
+    case spec
+    when :x11        then X11Requirement.new(spec.to_s, tags)
+    when :xcode      then XcodeRequirement.new(tags)
+    when :linux      then LinuxRequirement.new(tags)
+    when :macos      then MacOSRequirement.new(tags)
+    when :arch       then ArchRequirement.new(tags)
+    when :java       then JavaRequirement.new(tags)
+    when :osxfuse    then OsxfuseRequirement.new(tags)
+    when :tuntap     then TuntapRequirement.new(tags)
+    when :ld64       then ld64_dep_if_needed(tags)
+    else
+      opoo "Unsupported special dependency #{spec.inspect}" if ARGV.debug?
+    end
+  end
+
+  module Compat
+    def parse_string_spec(spec, tags)
+      opoo "'depends_on ... => :run' is disabled. There is no replacement." if tags.include?(:run) && ARGV.debug?
+      super
+    end
+  end
+
+  prepend Compat
+end
+
+module Homebrew
+  module_function
+
+  def extract
+    Homebrew::CLI::Parser.parse do
+      flag   "--tap="
+      flag   "--version="
+      switch :debug
+      switch :force
+    end
+
+    # If no formula args are given, ask specifically for a formula to be specified
+    raise FormulaUnspecifiedError if ARGV.named.empty?
+
+    # If some other number of args are given, provide generic usage information
+    raise UsageError if ARGV.named.length != 1
+
+    odie "The tap to which the formula is extracted must be specified!" if args.tap.nil?
```

Comment:
I wonder if it's worth making it a non-flagged argument (e.g. it's just the second argument)?

Reply:
I'll go in this direction. Just feels wrong making a flagged argument mandatory. In this case, we won't raise a separate `FormulaUnspecifiedError`, because there's no (easy) way to check if a missing argument is the missing formula or if it's the tap that needs to be specified. Instead, just a generic `UsageError` will be raised if we don't get exactly 2 named args.

Reply:
> Just feels wrong making a flagged argument mandatory

Reply:
> Just feels wrong making a flagged argument mandatory

Agreed.